modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EChain.java

```diff
@@ -233,6 +248,11 @@ private void analyzeWrite(final CompilerSettings settings, final Definition defi
     private void analyzeRead() {
         final ALink last = links.get(links.size() - 1);
 
+        // If the load node is a DEF node, we adapt its after type to use _this_ expected output type:
+        if (last instanceof ADefLink && this.expected != null) {
+            last.after = this.expected;
+        }
+
```

Comment:
If you want to give the cast way a shot I'm +1 on that.  It helps in the case of int int x; int y; x = y + 1.  When cast is called, we now know the expected type of the EChain, where during analyze we didn't because it hasn't been promoted yet.  But maybe this is too hacky?  Definitely a bit abusive of what casting is doing.  Just a thought to get the rest of the information out that we have.